core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
I think it remains to be seen whether having an unconstrained `M[_]` is worth the price of admission, but we'll find out. I also think it's likely that these three methods should be either two or one, even if performance may drive us away from it. The only method I see a need for is `Path => Metadata`. Paths which don't exist would return NoMetadata; the primary key is exposed through the metadata; the data is exposed through the metadata. This subsumes all manner of effects as well, because a metadata key need not produce `Data`, it could produce `M[Data]`.

I think this is true even after we have read/write filesystems. The interface through which you perform filesystem writes is (wait for it) available through the metadata.